pkg/packetfilter/iptables/iptables.go

```diff
@@ -0,0 +1,612 @@
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iptables
+
+import (
+	"crypto/sha256"
+	"encoding/base32"
+	"strconv"
+	"strings"
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/log"
+	"github.com/submariner-io/submariner/pkg/ipset"
+	"github.com/submariner-io/submariner/pkg/packetfilter"
+	utilexec "k8s.io/utils/exec"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	remoteCIDRIPSet                    = "SUBMARINER-REMOTECIDRS"
+	localCIDRIPSet                     = "SUBMARINER-LOCALCIDRS"
+	smPostRoutingChain                 = "SUBMARINER-POSTROUTING"
+	smGlobalnetEgressChainForPods      = "SM-GN-EGRESS-PODS"
+	smGlobalnetEgressChainForNamespace = "SM-GN-EGRESS-NS"
+	gnIPSetPrefix                      = "SM-GN-"
+	DriverName                         = "IPTables"
+)
+
+var (
+	tableTypeToStr             = [packetfilter.TableTypeMAX]string{"filter", "mangle", "nat"}
+	iphookChainTypeToStr       = [packetfilter.ChainTypeMAX]string{"filter", "mangle", "nat"}
+	iphookChainTypeToTableType = [packetfilter.ChainTypeMAX]packetfilter.TableType{
+		packetfilter.TableTypeFilter,
+		packetfilter.TableTypeRoute,
+		packetfilter.TableTypeNAT,
+	}
+	chainHookToStr  = [packetfilter.ChainHookMAX]string{"PREROUTING", "INPUT", "FORWARD", "OUTPUT", "POSTROUTING"}
+	ruleActiontoStr = [packetfilter.RuleActionMAX]string{"", "ACCEPT", "TCPMSS", "MARK", "SNAT", "DNAT"}
+	logger          = log.Logger{Logger: logf.Log.WithName(DriverName)}
+)
+
+type globalEgressSet struct {
+	ipSetNamed ipset.Named
+	snatCIDR   string
+	markValue  string
+}
+
+type IPTablesImpl struct {
+	Basic
+	remoteIPSet             ipset.Named
+	localIPSet              ipset.Named
+	stringToGlobalEgressSet map[packetfilter.GlobalEgressHandle]*globalEgressSet
+	IPSetIface              ipset.Interface
+}
+
+type Basic interface {
+	Append(table, chain string, rulespec ...string) error
+	AppendUnique(table, chain string, rulespec ...string) error
+	Delete(table, chain string, rulespec ...string) error
+	Insert(table, chain string, pos int, rulespec ...string) error
+	List(table, chain string) ([]string, error)
+	ListChains(table string) ([]string, error)
+	NewChain(table, chain string) error
+	ChainExists(table, chain string) (bool, error)
+	ClearChain(table, chain string) error
+	DeleteChain(table, chain string) error
+}
+
+func NewDriver() (packetfilter.Driver, error) {
```

Comment:
nit - "Driver" is redundant.

```suggestion
func New() (packetfilter.Driver, error) {
```